### PR TITLE
fix: prefer explicit dates in Tavily searches

### LINKED ISSUE #8154: [Feature] web_search_tavily conflicting time_range and date parameters always raises ERROR
### What happened / 发生了什么

When using the built-in `web_search_tavily` tool, if both `time_range` and `start_date` / `end_date` are passed, Tavily returns a 400 error.

The error message is:

```text
Tavily web search failed: {"detail":{"error":"When time_range is set, start_date or end_date cannot be set"}}, status: 400
```

This can happen easily because the tool schema exposes all of these optional parameters:
(不太聪明的模型可能会一直提供冲突参数)
- `time_range`
- `start_date`
- `end_date`

If the model fills both a relative range and explicit dates, AstrBot currently forwards all of them to Tavily without conflict cleanup.

### Reproduce / 如何复现？

Call `web_search_tavily` with parameters like:

```json
{
  "query": "example query info",
  "search_depth": "basic",
  "max_results": 5,
  "topic": "general",
  "time_range": "week",
  "start_date": "2026-05-10",
  "end_date": "2026-05-11"
}
```

Then Tavily returns:

```text
When time_range is set, start_date or end_date cannot be set
```

### Expected behavior / 期望行为

AstrBot should sanitize the payload before calling Tavily.

If `start_date` or `end_date` is provided, explicit date parameters should take priority, and `time_range` should be ignored.

### Affected code / 相关代码位置

File:

```text
astrbot/core/tools/web_search_tools.py
```

Current logic around `TavilyWebSearchTool.call()` builds the payload like this:

```python
time_range = kwargs.get("time_range", "")
if time_range in ["day", "week", "month", "year"]:
    payload["time_range"] = time_range
if kwargs.get("start_date"):
    payload["start_date"] = kwargs["start_date"]
if kwargs.get("end_date"):
    payload["end_date"] = kwargs["end_date"]
```

### Suggested fix / 建议修复

Prefer explicit dates over `time_range`:

```python
start_date = kwargs.get("start_date")
end_date = kwargs.get("end_date")
if start_date or end_date:
    # Tavily rejects requests that include both time_range and an explicit date.
    # Prefer explicit dates when both are provided by the model/user.
    if start_date:
        payload["start_date"] = start_date
    if end_date:
        payload["end_date"] = end_date
else:
    time_range = kwargs.get("time_range", "")
    if time_range in ["day", "week", "month", "year"]:
        payload["time_range"] = time_range
```

### Local verification / 本地验证

After applying the above patch locally, both of the following calls succeed:

#### 1. Date only

```json
{
  "query": "test query info",
  "search_depth": "basic",
  "max_results": 5,
  "topic": "general",
  "start_date": "2026-05-10",
  "end_date": "2026-05-11",
  "time_range": ""
}
```

#### 2. Both `time_range` and explicit dates

```json
{
  "query": "test query info",
  "search_depth": "basic",
  "max_results": 5,
  "topic": "general",
  "time_range": "week",
  "start_date": "2026-05-10",
  "end_date": "2026-05-11"
}
```

The second call no longer triggers Tavily's 400 error because `time_range` is not included in the final payload when dates are present.

### Environment / 环境

Observed in:

```text
AstrBot v4.24.2 from uv
Python 3.12
ubuntu 22.04
GPT-5.5
```

Stack trace excerpt:

```text
File ".../astrbot/core/tools/web_search_tools.py", line 450, in call
    results = await _tavily_search(provider_settings, payload)
File ".../astrbot/core/tools/web_search_tools.py", line 159, in _tavily_search
    raise Exception(
Exception: Tavily web search failed: {"detail":{"error":"When time_range is set, start_date or end_date cannot be set"}}, status: 400
```

### Use Case / 使用场景

_No response_

### Willing to Submit PR? / 是否愿意提交PR？

- [ ] Yes, I am willing to submit a PR. / 是的，我愿意提交 PR。

### Code of Conduct

- [x] I have read and agree to abide by the project's [Code of Conduct](https://docs.github.com/zh/site-policy/github-terms/github-community-code-of-conduct). /

### PR DESCRIPTION
**Pull request by Last-emo-boy**
Fixes #8154

### Modifications / 改动点

- Prefer explicit start/end dates when a Tavily request also contains time_range.
- Preserve the existing time_range behavior when no explicit date is provided.
- Add regression coverage for both dates, either single date, and empty dates.

- [x] This is NOT a breaking change. / 这不是一个破坏性变更。

### Screenshots or Test Results / 运行截图或测试结果

- uv run pytest -q tests/unit/test_web_search_tools.py: 27 passed
- ruff format --check .: passed
- ruff check .: passed

---

### Checklist / 检查清单

- [x] This change is covered by the linked issue.
- [x] My changes have been well-tested, and verification results are provided above.
- [x] No new dependencies are introduced.
- [x] My changes do not introduce malicious code.

## Summary by Sourcery

Normalize Tavily web search date filters to prefer explicit start/end dates over time_range while preserving existing behavior when no valid dates are provided.

Bug Fixes:
- Ensure Tavily searches use provided start/end dates instead of time_range when explicit non-empty dates are given.
- Fall back to time_range when start/end date inputs are empty or whitespace-only, avoiding invalid date payloads.

Tests:
- Add parametrized async tests covering Tavily date filter normalization for time_range-only, explicit dates, single-date, and empty/whitespace date cases.

### CODE REVIEW

File: astrbot/core/tools/web_search_tools.py [hunk 1]
```diff
@@ -653,13 +653,17 @@ async def call(self, context, **kwargs) -> ToolExecResult:
...
-            payload["start_date"] = kwargs["start_date"]
-        if kwargs.get("end_date"):
-            payload["end_date"] = kwargs["end_date"]
+        start_date = kwargs.get("start_date")
+        end_date = kwargs.get("end_date")
+        if start_date or end_date:
+            if start_date:
+                payload["start_date"] = start_date
+            if end_date:
+                payload["end_date"] = end_date
```
[the remark this answers is not in the document]
Comment (Last-emo-boy) [reply astrbot/core/tools/web_search_tools.py]: Addressed in add2ba899: both explicit date values are normalized with str(...).strip(), and a whitespace-only regression case was added.

**Comment by Last-emo-boy** [comment]
Reviewed: the allowed time_range list is not duplicated by this change; the existing check was moved unchanged into the mutually exclusive branch. Keeping this single-use validation inline follows the repository Inline-First rule.

**Review by Soulter — APPROVED** [review]
